src/d3a/d3a_core/simulation.py

```diff
@@ -444,20 +438,36 @@ def _handle_input(self, console, sleep: float = 0):
             if sleep == 0 or time.monotonic() - start >= sleep:
                 break
 
-    def _handle_paused(self, console):
-        if self.pause_after and self.time_since_start >= self.pause_after:
-            self.paused = True
-            self.pause_after = None
-        if self.paused:
-            start = time.monotonic()
-            log.critical("Simulation paused. Press 'p' to resume or resume from API.")
-            self.endpoint_buffer.update_stats(self.area, self.status, self.progress_info)
-            self.redis_connection.publish_intermediate_results(self.endpoint_buffer)
-            while self.paused:
+    def _handle_paused(self, console, tick_start):
+        if console is not None:
+            self._handle_input(console)
+            if self.pause_after and self.time_since_start >= self.pause_after:
+                self.paused = True
+                self.pause_after = None
+
+        started_pause = True
```

Comment:
Why do you need the started_pause here? You can add the logic that needs to happen once outside of the while loop, by having an additional check `if self.paused: ` , like the former implementation did. That way you are able to trigger things that need to happen only once before the loop starts .

Reply:
True! Will change this.

Reply:
https://github.com/gridsingularity/d3a/pull/779/commits/371dfdb5f89f0997df4df3aca50e248411256316